


<p>+++
title = “Projects”
description = “Development projects I have been involved in”
date = “2020-07-28”
aliases = [“about-us”,“about-hugo”,“contact”]
author = “Jakob Willforss”
+++</p>
<p>I have a range of programming interests that have touched on topics such as hackathons, Chinese study, game programming and more. I plan to gradually put up summaries of the different projects I have been involved in here.</p>
<div id="bioinformatics-software-development" class="section level1">
<h1>Bioinformatics software development</h1>
<p><strong>NormalyzerDE</strong> is a <a href="https://pubs.acs.org/doi/10.1021/acs.jproteome.8b00523">published</a> software for informed selection of normalization methods. It provides an ‘evaluation report’ guiding the user to select among a collection of normalization techniques. It provides a new approach to normalizing for electrospray intensity variations in the mass spectrometer. It has a built-in statistical module, which makes it easy to run the subsequent statistical step directly. It is available as a <a href="http://quantitativeproteomics.org/normalyzerde">web server</a> an <a href="https://github.com/ComputationalProteomics/NormalyzerDE">R package</a> and <a href="https://bioconductor.org/packages/release/bioc/html/NormalyzerDE.html">Bioconductor package</a>.</p>
<p><img src="/./projects_files/pr-2018-00523w_0007.jpeg" width="50" />
<img src="/projects_files/pr-2018-00523w_0007.jpeg" alt="NormalyzerDE" width="50%"/></p>
<p><strong>OmicLoupe</strong> is a software for rapid and comprehensive visualization of trends in omics-data. It provides interactive visualizations and introduces novel visualizations for understanding trends across multiple datasets, omics, or statistical comparisons. It is freely available as an <a href="https://github.com/ComputationalProteomics/OmicLoupe">open-source R package</a> and will be submitted for publication shortly.</p>
<p><img src="/projects_files/overview_visualization.png" alt="OmicLoupe" width="50%"/></p>
</div>
<div id="bioinformatics-analysis-projects" class="section level1">
<h1>Bioinformatics analysis projects</h1>
<p>During my Ph.D. I have been involved in three diverse agricultural projects, each using proteomics and, in some cases, transcriptomics to improve on the existing knowledge and breeding markers.</p>
<p><strong><em>Fusarium</em></strong> infection in oat: Proteogenomic exploration of the differing disease response between two oat varieties (<a href="https://pubmed.ncbi.nlm.nih.gov/32061841/">published</a>)</p>
<p><strong>Bull fertility</strong>: Proteomic exploration across four separate proteomic datasets taken from seminal plasma in bull (submitted).</p>
<p><strong>Potato latitude adaptability</strong>: Proteomic exploration across 17 potato-varieties grown 2016, and one variety is grown three Summers to find markers related to what makes some potatoes better handle the longer sun-hours and shorter growing seasons in northern Sweden (submitted shortly).</p>
</div>
<div id="game-development" class="section level1">
<h1>Game development</h1>
<p><strong>From Earth to Earth</strong> is a <a href="https://jakob37.itch.io/from-earth-to-earth-beta">freely available online</a> short game where a mud-man explores a rainy world searching for his Purpose. It is developed in C# and Unity. All code and assets are created by me. The source code is available <a href="https://github.com/Jakob37/FromEarthToEarth">here</a>.</p>
<p><img src="/projects_files/8CpHaL.png" alt="From Earth to Earth" width="50%"/></p>
<p><strong>The Sun Will Go Out</strong> is a <a href="http://www.linkpact.com/games/tswgo/index.html">freely available</a> space exploration and role-playing game developed over more than two years, in a joint effort between three developers. The source code is available <a href="https://github.com/LinkPact/The-Sun-Will-Go-Out">here</a>.</p>
<p><img src="/projects_files/18-highfence.png" alt="The Sun Will Go Out" width="50%"/></p>
</div>
